src/communication/domain_search_queue_policy.hpp

```diff
@@ -0,0 +1,80 @@
+#pragma once
+
+#ifndef NMC_USE_DOMAIN_SEARCH_QUEUE
+#error "domain_search_queue_policy.hpp should only be compiled in a NMC_USE_DOMAIN_SEARCH_QUEUE build"
+#endif
+
+#include <util/range.hpp>
+#include "base_communicator.hpp"
+
+namespace nest {
+namespace mc {
+namespace communication {
+
+using nest::mc::util::make_range;
+
+template <typename CommunicationPolicy>
+class domain_search_communicator: public base_communicator<CommunicationPolicy> {
+public:
+    using base = base_communicator<CommunicationPolicy>;
+    using typename base::event_queue;
+    using typename base::lt_spike_src;
+    using base::num_groups_local;
+    using base::base;
+
+protected:
+    using base::cell_group_index;
+    using base::connections_;
+
+public:
+    // go over each block of connections with the same source, and search for the
+    // associated block of spikes with the same source, and then push the product of events
+    // O(connection/spike * log(spikes/node))
+    std::vector<event_queue> make_event_queues(const gathered_vector<spike>& global_spikes)
+    {
+        // queues to return
+        auto queues = std::vector<event_queue>(num_groups_local());
+
+        // con_next is the beginning of the current block of connections with
+        // the same source
+        auto con_next = connections_.cbegin();
+        const auto con_end = connections_.cend();
+
+        // Search for next block of spikes and connections with the same sender
+        while (con_next != con_end) {
+            // we grab the next block of connections from the same sender
+            const auto src = con_next->source();
+            const auto domain = con_next->domain();
+            const auto targets = std::equal_range(con_next, con_end, src);
```

Comment:
Partly true. In fact, equal_range in g++ (and others I'd expect) does one binary search until it finds either the top or the bottom, then continues that binary search to find the other bound. So calling upper_bound is better, but not as much better as you'd expect. Should fix anyway -- it's confusing this way.

What I'm not sure of is, should we be doing a binary search at all, or should I do a walk forward to find the upper_bound? One is log(connections/node) the other length_of_run. Worst case the first is better, but how many connections do we really have on average to the same source?